File: android/app/src/main/java/com/masreferenceapp/network/NetworkUnencrypted.java
package com.masreferenceapp.network;

import android.net.DnsResolver;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.os.CancellationSignal;
import android.util.Log;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.common.SingleThreadAsserter;
import com.masreferenceapp.Constants;
import com.masreferenceapp.Status;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.HttpURLConnection;
import java.net.InetAddress;
import java.net.Socket;
import java.net.URL;
import java.net.UnknownHostException;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledThreadPoolExecutor;


public class NetworkUnencrypted extends ReactContextBaseJavaModule {
    ReactApplicationContext context;

    public NetworkUnencrypted(ReactApplicationContext context) {
        super(context);
        this.context = context;
    }

    @NonNull
    @Override
    public String getName() {
        return "NetworkUnencrypted";
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String resolveDns(){

        // Create a new instance of DnsResolver
        DnsResolver resolver = DnsResolver.getInstance();

        Executor mExecutor = Executors.newSingleThreadExecutor();
        CancellationSignal mcancellationSignal = new CancellationSignal();
        DnsResolver.Callback callback = new DnsResolver.Callback() {
            @Override
            public void onAnswer(@NonNull Object answer, int rcode) {
                System.out.println(answer.toString());
            }

            @Override
            public void onError(@NonNull DnsResolver.DnsException error) {

            }
        };

        // Resolve the domain name to its IP addresses
        resolver.query(null, Constants.remoteWebViewHttpDomain, DnsResolver.TYPE_A, DnsResolver.FLAG_NO_CACHE_LOOKUP, mExecutor, mcancellationSignal, callback);

        return Status.status("OK", resolver.toString());

    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String standardHTTP(){
        try{
            URL url = new URL("http://" + Constants.remoteWebViewHttpDomain);
            HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
            StringBuilder response = new StringBuilder();
            urlConnection.disconnect();
            return Status.status("OK", urlConnection.getResponseMessage());
        }
        catch (Exception e){
            return Status.status("FAIL", e.toString());
        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String nonStandardHTTP(){
        try{
            URL url = new URL("http", Constants.remoteWebViewHttpDomain, Constants.remoteWebViewHttpPort, "/");
            HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
            urlConnection.disconnect();
            return Status.status("OK", urlConnection.getResponseMessage());
        }
        catch (Exception e){
            return Status.status("FAIL", e.toString());
        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String rawTcp(){

        Socket socket;
        try{
            socket = new Socket(Constants.remoteWebViewHttpDomain, 80);
            // Get output stream to send data
            OutputStream outputStream = socket.getOutputStream();
            String request = "GET / HTTP/1.1\\nHost: " + Constants.remoteWebViewHttpDomain + "\\nX-A:msaTest\\n\\n";

            OutputStream output = socket.getOutputStream();
            PrintWriter writer = new PrintWriter(output, true);
            writer.print(request);
            socket.close();


            return Status.status("OK", socket.toString());

        }
        catch (Exception e){
            return Status.status("FAIL", "Failed to init socket.");


        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String rawUdp(){
        try {
            // Create a new UDP socket
            DatagramSocket socket = new DatagramSocket();

            InetAddress serverAddress = InetAddress.getByName(Constants.remoteWebViewHttpDomain);
            byte[] sendData = "HelloUDP".getBytes();
            DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, serverAddress, 53);

            // Send the UDP packet
            socket.send(sendPacket);
            socket.close();
            return Status.status("OK",  socket.toString());
        } catch (Exception e) {
            return Status.status("FAIL", "Failed to init socket.");
        }
    }


}

File: android/app/src/main/java/com/masreferenceapp/network/NetworkUnencrypted.kt
package com.masreferenceapp.network

import android.webkit.WebView
import android.webkit.WebViewClient
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.masreferenceapp.MasSettings
import com.masreferenceapp.ReturnStatus
import com.masreferenceapp.network.helpers.WebViewJavaScriptBridge
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.HttpURLConnection
import java.net.InetAddress
import java.net.Socket
import java.net.URL


class NetworkUnencrypted(var context: ReactApplicationContext) : ReactContextBaseJavaModule(
    context
) {
    override fun getName(): String {
        return "NetworkUnencrypted"
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun openHTTP(): String {
        val testDomain = MasSettings.getTestDomain()
        val r = ReturnStatus()

        try {
            val url = URL("http://$testDomain")
            val connection = url.openConnection() as HttpURLConnection
            connection.connectTimeout = 500
            connection.readTimeout = 500
            connection.connect()
            r.success("Pin verification successful. Connection to https://$testDomain established. Response code was: ${connection.responseCode}.")
            connection.disconnect()
        } catch (e: Exception) {
            r.fail(e.toString())
        }
        return r.toJsonString()
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun nonStandardHTTP(): String {
        val testDomain = MasSettings.getTestDomain()
        val port = 3001

        val r = ReturnStatus()

        try {
            val url = URL("http", testDomain, port, "/")
            val connection = url.openConnection() as HttpURLConnection
            connection.connectTimeout = 500
            connection.readTimeout = 500

            connection.connect()
            r.success("Pin verification successful. Connection to http://$testDomain:$port established. Response code was: ${connection.responseCode}.")
            connection.disconnect()
        } catch (e: Exception) {
            r.fail(e.toString())
        }
        return r.toJsonString()
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun rawTcp(): String {
        val testDomain = MasSettings.getTestDomain()
        val port = 5001
        val r = ReturnStatus()

        val socket: Socket
        try {
            socket = Socket(testDomain, port)
            socket.soTimeout = 500

            val request = "This is the MAS app using TCP."
            val output = socket.getOutputStream()
            val writer = PrintWriter(output)
            writer.print(request)
            writer.flush()

            //receive the message which the server sends back
            val input = BufferedReader(InputStreamReader(socket.getInputStream()))
            val message = input.readLine();

            r.success("Raw TCP socket to $testDomain:$port successfully established. Received data: $message")

            socket.close()


        } catch (e: Exception) {
            r.fail(e.toString())
        }
        return r.toJsonString()
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun rawUdp(): String {
        val testDomain = MasSettings.getTestDomain()
        val port = 4001
        val r = ReturnStatus()

         try {

             val socket = DatagramSocket()
             socket.soTimeout = 500
             val serverAddress = InetAddress.getByName(testDomain)
             val sendData = "This is the MAS app using UDP.".toByteArray()
             val sendPacket = DatagramPacket(sendData, sendData.size, serverAddress, port)

             // Send the UDP packet
             socket.send(sendPacket)


             val buffer = ByteArray(512)
             val packet = DatagramPacket(buffer, buffer.size)
             socket.receive(packet);
             val receivedData = String(packet.data, 0, packet.length)

             socket.close()

             r.success("Raw UDP socket to $testDomain:$port successfully established. Received data: $receivedData")
         } catch (e: Exception) {
             r.fail(e.toString())
         }
        return r.toJsonString()
    }


    @ReactMethod()
    fun webSocket(promise: Promise) {
        val r = ReturnStatus()
        val testDomain = "ws://"+MasSettings.getTestDomain()+":2001"

        val wv = WebView(context)
        wv.settings.javaScriptEnabled = true
        wv.webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView, url: String) {
                view.evaluateJavascript("initWebSocket('$testDomain')") {
                    r.success("WebView successfully loaded.")
                }
            }
        }

        val jsBridge = WebViewJavaScriptBridge(testDomain, promise, r)
        wv.addJavascriptInterface(jsBridge, "javaScriptBridge")
        wv.loadUrl("file:///android_asset/ws.html")
    }

    //@method

}